STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_registry.c

```diff
@@ -0,0 +1,300 @@
+
+  az_result result;
+  az_ulib_registry_node* runner;
+
+  /* Find node containing key */
+  runner = (az_ulib_registry_node*)(&__REGISTRYINFO);
+  for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+  {
+    if(runner->_internal.status.delete_flag != REGISTRY_DELETED)
+    {
+      /* If the content is not deleted and ready for use */
+      if (runner->_internal.status.ready_flag == REGISTRY_READY)
+      {
+        if (az_span_is_content_equal(key, runner->_internal.data.key))
+        {
+          *value = runner->_internal.data.value;
+          result = AZ_OK;
+          break;
+        }
+      }
+      else 
+      {
+        result = AZ_ERROR_ITEM_NOT_FOUND;
+        break;
+      }
+    }
+    runner++;
+  }
+
+  return result;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_add(az_span key, az_span value)
+{
+  /* Precondition check */
+  _az_PRECONDITION_VALID_SPAN(key, 1, false);
+  _az_PRECONDITION_VALID_SPAN(value, 1, false);
+
+  AZ_ULIB_TRY
+  {
+    az_span sample_value;
+    az_result result;
+    az_ulib_registry_node new_node;
+    uint8_t* new_node_ptr;
+    uint8_t* key_dest_ptr;
+    uint8_t* val_dest_ptr;
+
+    az_pal_os_lock_acquire(&registry_lock);
+    {
+      /* Validate for duplicates before adding new entry */
+      result = az_ulib_registry_get(key, &sample_value);
+      AZ_ULIB_THROW_IF_ERROR((result == AZ_ERROR_ITEM_NOT_FOUND), AZ_ERROR_ULIB_ELEMENT_DUPLICATE);
+
+      /* Find destination in flash buffer */
+      key_dest_ptr = flash_buf_get();
+      val_dest_ptr
+          = (uint8_t*)((int32_t)key_dest_ptr + ((((az_span_size(key) - 1) >> 3) + 1) << 3));
+
+      /* Handle out of space scenario */
+      AZ_ULIB_THROW_IF_ERROR(
+          (key_dest_ptr < ((uint8_t*)(&__REGISTRY) + MAX_AZ_ULIB_REGISTRY_BUFFER)),
+          AZ_ERROR_OUT_OF_MEMORY);
+      AZ_ULIB_THROW_IF_ERROR(
+          (val_dest_ptr < ((uint8_t*)(&__REGISTRY) + MAX_AZ_ULIB_REGISTRY_BUFFER)),
+          AZ_ERROR_OUT_OF_MEMORY);
+
+      /* Set free node information */
+      new_node._internal.data.key = az_span_create(key_dest_ptr, az_span_size(key));
+      new_node._internal.data.value = az_span_create(val_dest_ptr, az_span_size(value));
+
+      /* Update registry node in flash */
+      new_node_ptr = NULL;
+      AZ_ULIB_THROW_IF_AZ_ERROR(store_registry_node(new_node, &new_node_ptr));
+
+      /* Write key to flash */
+      AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(
+          internal_flash_write(key_dest_ptr, az_span_ptr(key), az_span_size(key))));
+
+      /* Write value to flash */
+      AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(
+          internal_flash_write(val_dest_ptr, az_span_ptr(value), az_span_size(value))));
+
+      /* After successful storage of registry node and actual key value pair, set flag in node to
+      indicate the entry is now ready to use  */
+      AZ_ULIB_THROW_IF_AZ_ERROR(set_registry_node_ready_flag(new_node_ptr));
+    }
+    az_pal_os_lock_release(&registry_lock);
```

Comment:
>az_pal_os_lock_release [](http://example.com/codeflow?start=4&length=22)

If any of the calls inside of the lock thow an error, this release will not run, and you will lock the system forever. So please move the entire `TRY/CATCH` inside of the lock.